### PR TITLE
[2.11.x] DDF-3185 Config admin plugins now receive config state which greatly simplifies locking semantics

### PR DESCRIPTION
**Pull request by Lambeaux**
#### What does this PR do?
Changes `ConfigurationPersistencePlugin` to be initializable with a set of configs representing the current configuration state of the system. This continues to ensure no race conditions occur while Karaf starts and, to that same point, encapsulates the locking in a single place. 

#### Who is reviewing it? 
@brjeter 
@paouelle 

#### Choose 2 committers to review/merge the PR. 
@coyotesqrl 
@lessarderic

#### How should this be tested? (List steps with links to updated documentation)
See testing steps on https://github.com/codice/ddf/pull/2643

#### Any background context you want to provide?
DDF now embeds and wraps Felix config admin to provide additional customization. Prior to this PR, a lock was being passed around to account for Felix's configuration caching layer. This PR simplifies that so it's no longer necessary to pass around a lock. 

#### What are the relevant tickets?
[DDF-3185](https://codice.atlassian.net/browse/DDF-3185)

#### Notes on Review Process
For the general case, merging is gated on the following:

* Review by at least two contributors
* Review by at least two committers
* Successful CI
* Successful static analysis (SonarQube/Coverity/etc.)
* Heroing of change (details vary by ticket)

Certain exceptions may occur. For example, if there is a problem with the CI environment then a manually performed full build and static analysis check may be substituted _at the discretion of the repository maintainers_. Additionally, repository metadata such as this `PULL_REQUEST_TEMPLATE`, the repository `README`, and other non-functional items may be merged without requiring CI and static analysis.

Finally, an abbreviated review process may be in place for forward-port tickets immediately following branching and prior to branch divergence. In that case, approvals, CI, static analysis, and heroing that occur against the PR against the _release branch_ can be considered sufficient for reviewing the PR against the _master branch_.

Once the branches diverge, the process will return to normal.

Insofar as divergence is not a function of time but of change to the parent branch, the repository maintainers will determine when abbreviated review ends.

When abbreviated review is in effect, a note to that effect will be at the top of this template. Following is an example note:
```
##### ABBREVIATED REVIEW BETWEEN 2.12.X AND MASTER IS IN EFFECT
____

```

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java [hunk 1]
```diff
@@ -107,40 +104,23 @@ public ConfigurationUpdater(
...
-            });
+  public void initialize(List<ConfigurationContext> state) {
+    state.forEach(
+        context -> {
+          try {
+            handleStore(context);
+          } catch (IOException e) {
+            LOGGER.error(
+                "Problem updating config file [{}]. {}", context.getConfigFile(), e.getMessage());
+            if (LOGGER.isDebugEnabled()) {
```
Comment (paouelle) [inline platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java]: ❓ Why the if(), you should be fine to simply call debug()
But this can be done in one call to error() by using the error(String, Throwable) version

File: platform/osgi/platform-osgi-internal-api/src/main/java/org/codice/felix/cm/internal/ConfigurationPersistencePlugin.java [hunk 2]
```diff
@@ -39,10 +39,9 @@
    * available to be called, unlike an OSGi life-cycle {@code init()} method, which is called prior
    * to the service becoming available.
    *
-   * @param factory the configuration context factory used for obtaining {@link
-   *     ConfigurationContext}s.
+   * @param state the current configuration state of the system.
    */
-  void initialize(ConfigurationContextFactory factory);
+  void initialize(List<ConfigurationContext> state);
```
Comment (paouelle) [inline platform/osgi/platform-osgi-internal-api/src/main/java/org/codice/felix/cm/internal/ConfigurationPersistencePlugin.java]: ❓ Should this be a set to indicate there will be no duplicates?

File: platform/admin/core/admin-core-impl/src/main/java/org/codice/ddf/admin/core/impl/ConfigurationAdminImpl.java [hunk 3]
```diff
@@ -161,11 +161,6 @@ private Bundle getBoundBundle(Configuration config) {
     return null;
   }
 
-  public Configuration[] listConfigurations(String filter)
```
Comment (paouelle) [inline platform/admin/core/admin-core-impl/src/main/java/org/codice/ddf/admin/core/impl/ConfigurationAdminImpl.java]: 👍

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigPluginCustomizer.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ConfigPluginCustomizer} is where plugin initialization occurs.
+ *
+ * <p>Plugin initialization delivers the list of all known configurations to the plugins so they can
+ * prepare for future updates and deletions. The {@code servieStartingLock} ensures no additional
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigPluginCustomizer.java]: ✏️  {@code serviceStartingLock}

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java [hunk 5]
```diff
@@ -20,33 +20,37 @@
...
+  private final StampedLock serviceStartingLock = new StampedLock();
+
+  public DelegatingPersistenceManager(PersistenceManager persistenceManager) {
+    super(persistenceManager);
+    this.configPersistenceTracker =
         new ServiceTracker<>(
             getBundleContext(),
             ConfigurationPersistencePlugin.class,
-            new ConfigInterceptingCustomizer(configDeadZoneLock)));
+            new ConfigPluginCustomizer(getInnerPersistenceManager(), serviceStartingLock));
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: ❗️ Could we avoid passing the lock outside of this class and keep everything internal instead? 

If not then I would really suggest not exposing what type of lock it is. The plugin customizer class doesn't need to know why it has to grab a write lock instead of a read lock. This is very specific to the implementation if the DelegatingPersistenceManager class.

My preference would be to keep that class responsible for detecting new plugins and call this delegating persistence manager when it does such that this persistence manager is responsible for all config objects related operations and in this case, it would be to grab the lock, get all configs from the inner persistence manager and call the init for that new service added. This keeps all interactions with the services listening for events and all cache interactions (read, update, delete) in one place; here.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java [hunk 6]
```diff
@@ -65,25 +69,37 @@ public void close() throws Exception {
 
   @Override
   public void store(String pid, Dictionary properties) throws IOException {
-    ConfigurationContextImpl context = createContext(pid, properties);
-    if (context.shouldBeVisibleToPlugins()) {
-      List<ConfigurationPersistencePlugin> plugins =
-          asPlugins(configPersistenceTracker.getServices());
-      for (ConfigurationPersistencePlugin plugin : plugins) {
-        plugin.handleStore(context);
+    // Read lock because Felix provides its own protection for multiple config deletes
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: ❓ Do we need that comment? Or are we locking just to make sure that an initialization is not currently happening? Same in the other methods

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java [hunk 7]
```diff
@@ -65,25 +69,37 @@ public void close() throws Exception {
...
-    super.store(pid, properties);
   }
 
   @Override
   public void delete(String pid) throws IOException {
-    List<ConfigurationPersistencePlugin> plugins =
-        asPlugins(configPersistenceTracker.getServices());
-    for (ConfigurationPersistencePlugin plugin : plugins) {
-      plugin.handleDelete(pid);
+    // Read lock because Felix provides its own protection for multiple config deletes
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: see above.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/WrappedPersistenceManager.java [hunk 8]
```diff
@@ -36,6 +36,10 @@
     this.persistenceManager = persistenceManager;
   }
 
+  PersistenceManager getInnerPersistenceManager() {
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/WrappedPersistenceManager.java]: ❓ Should this be protected since it is meant for subclasses? 
If it is package-private because of testing, then annotate with @VisibleForTesting

File: platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java (hunk 1 above)
Comment (Lambeaux) [reply platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java]: I'm trying to discourage the trend of throwing stack traces in error messages. Those should be reserved for debug. We do this everywhere and it makes the logs ugly.

**Comment by Lambeaux** [comment]
Build now

File: platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java (hunk 1 above)
Comment (paouelle) [reply platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java]: 👍 for the debug but you still do not need the if before it

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java [hunk 9]
```diff
@@ -147,6 +147,27 @@ public boolean shouldBeVisibleToPlugins() {
     return servicePid != null && configIsNew == null && pidList == null && propertyCount > 0;
   }
 
+  // Contexts can be used in a set
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: ✏️  This can be replaced with `if (!getClass().isInstance(o)) {` but usually, one simply checks if `o` is not an instance of the current class (i.e. `ConfigurationContextImpl`) as in `if (!(o instanceof ConfigurationContextImpl)) {`

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java [hunk 10]
```diff
@@ -147,6 +147,27 @@ public boolean shouldBeVisibleToPlugins() {
...
+      return false;
+    }
+
+    ConfigurationContextImpl context = (ConfigurationContextImpl) o;
+
+    return servicePid != null ? servicePid.equals(context.servicePid) : context.servicePid == null;
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: ✏️  Can be replace with `return Objects.equals(servicePid, context.servicePid);`

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java [hunk 11]
```diff
@@ -147,6 +147,27 @@ public boolean shouldBeVisibleToPlugins() {
...
+  }
+
+  // Contexts can be used in a set
+  @Override
+  public int hashCode() {
+    return servicePid != null ? servicePid.hashCode() : 0;
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: ✏️  Can be replaced with `return Objects.hash(servicePid);`

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java [hunk 12]
```diff
@@ -13,40 +13,73 @@
...
-  public DelegatingPersistenceManager(
-      PersistenceManager persistenceManager, StampedLock configDeadZoneLock) {
-    this(
-        persistenceManager,
+  private final StampedLock serviceStartingLock = new StampedLock();
+
+  public DelegatingPersistenceManager(PersistenceManager persistenceManager) {
+    super(persistenceManager);
+    this.configPersistenceTracker =
+        // Passing this as the customizer instead of passing a lock between objects
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: ✏️  Not sure we need this comment anymore.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java [hunk 13]
```diff
@@ -65,32 +98,85 @@ public void close() throws Exception {
...
+        ConfigurationContextImpl context =
+            new ConfigurationContextImpl((String) props.get(SERVICE_PID), props);
+        if (context.shouldBeVisibleToPlugins()) {
+          configs.add(context);
+        }
+      }
+      plugin.initialize(configs);
+    } catch (IOException e) {
+      LOGGER.error(
+          format(
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: If you are not passing the exception to error(), then use its standard `error(String, Object...)` form instead of calling `format()`. 
It would be nice to trace that stack trace though!

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/WrappedPersistenceManager.java [hunk 14]
```diff
@@ -36,6 +36,10 @@
     this.persistenceManager = persistenceManager;
   }
 
+  protected PersistenceManager getInnerPersistenceManager() {
```
Comment (paouelle) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/WrappedPersistenceManager.java]: ❗️ Can be private now or simply removed since the calling code is a method of this class now.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java (hunk 10 above)
Comment (paouelle) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: ❓ Can servicePid actually be null?

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java [hunk 15]
```diff
@@ -108,28 +194,17 @@ private static BundleContext getBundleContext() {
...
-      ConfigurationPersistencePlugin plugin = getBundleContext().getService(serviceReference);
-      // Write lock because no configs should be getting saved or deleted while we initialize
-      long stamp = configDeadZoneLock.writeLock();
-      try {
-        plugin.initialize(factory);
-      } finally {
-        configDeadZoneLock.unlockWrite(stamp);
-      }
-      return plugin;
+      return handleServiceAdded(serviceReference);
```
Comment (brjeter) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: ✏️  At this point, does it not make sense to just have `DelegatingPersistenceManager` implement the `ServiceTrackerCustomizer` interface? I'm saying that because this inner class literally just delegates to a method on the outer class.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java [hunk 16]
```diff
@@ -65,32 +98,85 @@ public void close() throws Exception {
...
+  }
+
+  private ConfigurationPersistencePlugin handleServiceAdded(
+      ServiceReference<ConfigurationPersistencePlugin> serviceReference) {
+    ConfigurationPersistencePlugin plugin = retrieveServiceObject(serviceReference);
+    // Write lock because no configs should be getting saved or deleted while we initialize
+    long stamp = serviceStartingLock.writeLock();
+    try {
+      Enumeration<Dictionary<String, Object>> dictionaries =
+          getInnerPersistenceManager().getDictionaries();
```
Comment (brjeter) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: It should be noted that this is a very expensive operation. I think it's fine realistically because there will never be several `ConfigurationPersistencePlugin`s. But it may be worth making a tech debt ticket to address it.

Comment (paouelle) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: Not sure it should be addressed any other ways. The idea is to avoid going back to cfg admin which can lead to deadlock issues or lock complications. Since the persistent manager already manages all updates, relying on its own cache (i.e. the data/cache directory) is clean. To optimize this, we would need to create yet another cache and in this case, as you mentioned, there will never be many of these and it will typically happen only at config time. As such, this implementation is more than sufficient and I don't think we should create a tech debt ticket.

Comment (brjeter) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: I agree that I cannot think of a better solution. I just wanted to mention it. It may be worth noting somewhere in the javadoc at least then. Maybe on the `DelegatingPersistenceManager` as well as the `ConfigurationPersistencePlugin` interface?

Comment (brjeter) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: Just to warn future developers in this code that adding a `ConfigurationPersistencePlugin` should not be done lightly.

Comment (paouelle) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: 👍  on adding a comment in the code

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java (hunk 9 above)
Comment (Lambeaux) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: We should still keep the null protection. Changing the rest. 👍

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java (hunk 10 above)
Comment (Lambeaux) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: Technically no, according to the contract and where I'm getting it from, it should not be null. But I'm not rolling the dice with our inspection of Felix sources.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java (hunk 13 above)
Comment (Lambeaux) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: Will debug the stack trace. Changed the error log to use it's default interpolation technique.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/WrappedPersistenceManager.java (hunk 14 above)
Comment (Lambeaux) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/WrappedPersistenceManager.java]: This is not entirely correct unless you assumed we would make the **field** it returns protected or package private? This method is called in a child class. We still need protected or package-private access on this method.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java (hunk 15 above)
Comment (Lambeaux) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: Originally that's how I had it, but it requires the pass of the `this` pointer in the constructor of delegating persistence manager. If an exception gets thrown we could have a floating reference to a partially constructed object.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java (hunk 16 above)
Comment (Lambeaux) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: Even if we had 12 config persistence plugins, that would translate to (at worst) an extra 12 - 24 seconds of container boot time (assuming it takes 1 -2 seconds to read the entire cache from disk). This only impacts plugin boot time, not plugin performance time in the heat of production. I'll provide background context, but I think this is trivial. 

Comments added inline and in javadoc.

Comment (Lambeaux) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: When I open the discussion with Felix config admin maintainers about possibly merging our solutions / framework improvements, I'll be sure to mention interoperability with a proper caching solution, such as Apache Ignite.

**Comment by Lambeaux** [comment]
Comments addressed. Build now.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java (hunk 9 above)
Comment (paouelle) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: `null instanceof <something>` will always be `false`
As for subclasses, it won't break. It will verify the base class is correct which is what you want. The subclass would typically be calling super.equals() and then test for its own instance before verifying its own attributes.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java (hunk 10 above)
Comment (paouelle) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: Then my first comment applies here.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/WrappedPersistenceManager.java (hunk 14 above)
Comment (paouelle) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/WrappedPersistenceManager.java]: Sorry, I thought it was only called in this class.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java (hunk 16 above)
Comment (paouelle) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: @Lambeaux If you had 12 persistence plugin, then I would suggest not having a serial lock. Since the lock is designed to prevent updates while you take a snapshot, you could easily have the second persistence plugin starts its initialization at the same time. So detecting the lock is already acquired and actually wait for the same caching result to initialize as opposed to reloading it. Thus you would still have a 1-2 seconds delay unless it happens after the first one is done initializing. There is ways to even optimize this by temporarily caching stuff instead of doing it throughout runtime. There is a point where new persistence plugins should no longer be initialized and caching would not be required. In any case, this is moot right now so just a comment that if it ever happens, we should do something about it or better, a warning generated if a second plugin is registered might be something nice to get in the logs.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java [hunk 17]
```diff
@@ -65,32 +97,86 @@ public void close() throws Exception {
...
 
+  /* Factory method visible for testing purposes - to inject a mock for the plugin */
+  ConfigurationPersistencePlugin retrieveServiceObject(
+      ServiceReference<ConfigurationPersistencePlugin> serviceReference) {
+    return FrameworkUtil.getBundle(DelegatingPersistenceManager.class)
+        .getBundleContext()
+        .getService(serviceReference);
+  }
+
+  private ConfigurationPersistencePlugin handleServiceAdded(
```
[the remark this answers is not in the document]
Comment (paouelle) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java]: Time to suppress this one. It is better to have it here then in the subclass.

File: platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/admin/core/api/ConfigurationAdmin.java [hunk 18]
```diff
@@ -80,13 +79,6 @@
    */
   List<Service> listServices(String serviceFactoryFilter, String serviceFilter);
 
-  /**
-   * Returns all configurations that satisfy the provided filter.
-   *
-   * @see org.osgi.service.cm.ConfigurationAdmin#listConfigurations(String)
-   */
-  Configuration[] listConfigurations(String filter) throws IOException, InvalidSyntaxException;
```
Comment (stustison) [inline platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/admin/core/api/ConfigurationAdmin.java]: Why was this removed?

Comment (Lambeaux) [reply platform/admin/core/admin-core-api/src/main/java/org/codice/ddf/admin/core/api/ConfigurationAdmin.java]: It was only added on the last DDF-3185 PR because the plugin needed it. When we changed the design, it was no longer needed.

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java (hunk 9 above)
Comment (Lambeaux) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/ConfigurationContextImpl.java]: I still think using `instanceof` would violate the least astonishment principle. An instance of a different, albeit related, class should not evaluate as equal.

**Comment by Lambeaux** [comment]
Rebuilding after adding a Sonar suppression. Should still be passing. 
Build now.

**Comment by idperez** [comment]
hero'd on mac

**Comment by AzGoalie** [comment]
successfully heroed on Windows 👍
